Allow error handler to log stream errors

nwoltman opened the pull request:
This lets custom error handlers have a chance to handle the error (mainly to log it) which fixes a bug where stream errors are completely lost if the built-in logger is not turned on.

This change also makes Fastify log stream errors with all of the information of any other error.

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows [Code of conduct](https://github.com/fastify/fastify/blob/master/CODE_OF_CONDUCT.md)

allevo commented on lib/reply.js:
```diff
@@ -216,6 +211,11 @@ function handleError (reply, error, cb) {
   }
 
+  if (reply.res.headersSent) {
+    reply.res.log.info('response terminated with an error with headers already sent')
```
reply.requesy.log is nicer imho

nwoltman replied:
I would change it, but `reply.res.log` is used just 20 lines above this one so it would look very strange to log in different ways in the same function.

mcollina replied:
can you put `log` into its own variable then?

nwoltman replied:
Done 👍

allevo commented on test/stream.test.js:
```diff
@@ -168,3 +168,54 @@ test('Destroying streams prematurely', t => {
+  const http = require('http')
+
+  fastify.setErrorHandler(function (err, reply) {
+    t.ok(err)
```
Can you add a check for errors.code here?

nwoltman replied:
`err.code` is `undefined` 🤔 
I changed it to `t.type(err, Error)` though.

mcollina commented on test/stream.test.js:
```diff
@@ -168,3 +168,54 @@ test('Destroying streams prematurely', t => {
+  fastify.setErrorHandler(function (err, reply) {
+    t.ok(err)
+    try {
+      reply.send('error message')
```
this should be an `Error`  object.

nwoltman replied:
I added another try-catch that tries to send an `Error` object. Although it doesn't really matter what gets sent because they both end up here with `this.sent` being `true`:
https://github.com/fastify/fastify/blob/a94ffdc2be92300f2bfa566a80c8dc6cdc39c6de/lib/reply.js#L31-L35

mcollina replied:
Oh yes, my bad.

delvedor commented on docs/Server-Methods.md:
````diff
@@ -138,3 +138,14 @@ Set the schema compiler for all routes [here](https://github.com/fastify/fastify
+
+```js
+fastify.setErrorHandler(function (error, reply) {
+  if (reply.res.headersSent) {
````
Should we perform this check internally and not call the custom error handler at all?

nwoltman replied:
That's what this change is all about. The custom error handler *should* be called so that users have a chance to log the error in their own way (not everyone has Fastify's logger enabled).

delvedor replied:
Ops, I misunderstood then.
I don't know if I'm really ok with this change, this could be really annoying.
Since there is the chance to use an external logger I don't see why a user should write some complex code to integrate an external logger that is not integrated within the framework.
I'd like to have the opinion of the other on this point.

mcollina replied:
I'm not ok with this change either.
Given the logger is pluggable, I don't see much of a reason to make the user aware of this for only logging purposes.

nwoltman replied:
It's not really about integrating a logger using Fastify's `logger` option. It's more that not everyone agrees with the way that things are logged. Maybe I want some [400 errors to be logged](https://github.com/fastify/fastify/blob/629c16c9c9d09109a264f8868f7f0844907bb5a3/lib/reply.js#L198-L199) as `warn` instead of `info`, or to log more information in addition to `res` and `err`. Maybe I want to log stream errors differently than other errors.

Some people switch to Fastify from something else and already have specific log messages and analytics for those messages. Not giving users the chance to log things in their own way (or at all) isn't very nice. 

I agree this change seems wrong if you see the custom error handler as a way to send a response like a route handler, but I see the custom error handler as a way to handle the error in my own way.

mcollina replied:
A solution of this might be to make `res.send()` idempotent if the headers are already been sent. However, this might cause more damage than what we are trying to solve, because it can hide some bugs.

The error handler is for errors that can be acted upon. That's why it gets a `reply` object. Errors that cannot get acted upon should only be logged. BTW, I'm fine if we log with warn level.

If you want to handle this case differently, you can do a lot of other things. Sorry.

nwoltman commented:
Continuing from @mcollina's [comment](https://github.com/fastify/fastify/pull/634#discussion_r160215208).

> A solution of this might be to make res.send() idempotent if the headers are already been sent.

That's essentially what this PR does. `reply.sent` gets set to `true` if `res.headersSent` is true so calling `reply.send()` will do nothing (other than log that it was called when the reply was already sent). That's why I made the documentation say "if you want to avoid any unnecessary sending logic". They don't actually need to do the `if (reply.sent)` check (I updated the docs to this just now), it's just for performance.

> The error handler is for errors that can be acted upon.

I guess this brings up the idea that different people think of "acting upon" an error in different ways. I would act upon the error by logging it and responding if possible. Maybe someone else wants to do something other than log or respond. It's just nice to have the freedom to choose.

@delvedor @mcollina I updated the docs and tests to make these changes (hopefully) easier to approve. But if you don't approve I can undo all these changes and just change the logging [here](https://github.com/fastify/fastify/blob/629c16c9c9d09109a264f8868f7f0844907bb5a3/lib/reply.js#L176) a little. I can always configure Fastify in my app with a logger that only logs errors if headers are already sent.

nwoltman commented:
I just updated the documentation due to the fact that setting headers after headers are already sent will throw an error.

I think I'd be okay with closing this PR if this:

https://github.com/fastify/fastify/blob/629c16c9c9d09109a264f8868f7f0844907bb5a3/lib/reply.js#L176

was changed to something like this:

```js
reply.res.log.error(err, 'response terminated with an error with headers already sent')
// or `log.warn`
```

jsumners commented:
I'd say that line should be logged at the `error` level to begin with.

nwoltman commented:
Closed by #637
